File: app/src/main/java/com/ceaver/assin/database/converter/Converters.java
package com.ceaver.assin.database.converter;

import android.arch.persistence.room.TypeConverter;

import com.ceaver.assin.alerts.AlertType;
import com.ceaver.assin.assets.AssetCategory;
import com.ceaver.assin.intentions.IntentionStatus;
import com.ceaver.assin.intentions.IntentionType;
import com.ceaver.assin.markets.Title;
import com.ceaver.assin.markets.TitleRepository;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.Optional;
import java.util.UUID;

public class Converters {

    @TypeConverter
    public static LocalDate daysFromEpochToLocalDate(Long daysFromEpoch) {
        return daysFromEpoch == null ? null : LocalDate.MIN.plusDays(daysFromEpoch);
    }

    @TypeConverter
    public static Long localDateToDaysFromEpoch(LocalDate date) {
        return date == null ? null : ChronoUnit.DAYS.between(LocalDate.MIN, date);
    }

    @TypeConverter
    public static LocalDateTime toLocalDateTime(Long secondsFromEpoch) {
        return secondsFromEpoch == null ? null : LocalDateTime.MIN.plusSeconds(secondsFromEpoch);
    }

    @TypeConverter
    public static Long fromLocalDateTime(LocalDateTime dateTime) {
        return dateTime == null ? null : ChronoUnit.SECONDS.between(LocalDateTime.MIN, dateTime);
    }

    @TypeConverter
    public static AlertType toAlertType(String name) {
        return AlertType.valueOf(name);
    }

    @TypeConverter
    public static String fromAlertType(AlertType alertType) {
        return alertType.name();
    }

    @TypeConverter
    public static UUID toUuid(String uuid) {
        return UUID.fromString(uuid);
    }

    @TypeConverter
    public static String fromUuid(UUID uuid) {
        return uuid.toString();
    }

    @TypeConverter
    public static Optional<Double> toOptionalOfDouble(String optionalDouble) {
        return optionalDouble.isEmpty() ? Optional.empty() : Optional.of(Double.parseDouble(optionalDouble));
    }

    @TypeConverter
    public static String fromOptionalOfDouble(Optional<Double> optionalDouble) {
        return optionalDouble.isPresent() ? optionalDouble.get().toString() : "";
    }

    @TypeConverter
    public static Optional<LocalDateTime> toOptionalLocalDateTime(Long secondsFromEpoch) {
        return secondsFromEpoch == -1 ? Optional.empty() : Optional.of(LocalDateTime.MIN.plusSeconds(secondsFromEpoch));
    }

    @TypeConverter
    public static Long fromOptionalLocalDateTime(Optional<LocalDateTime> optionalLocalDateTime) {
        return optionalLocalDateTime.isPresent() ? ChronoUnit.SECONDS.between(LocalDateTime.MIN, optionalLocalDateTime.get()) : -1;
    }

    @TypeConverter
    public static AssetCategory toAssetCategory(String name) {
        return AssetCategory.valueOf(name);
    }

    @TypeConverter
    public static String fromAssetCategory(AssetCategory assetCategory) {
        return assetCategory.name();
    }

    @TypeConverter
    public static Optional<String> toOptionalString(String string) {
        return string.isEmpty() ? Optional.empty() : Optional.of(string);
    }

    @TypeConverter
    public static String fromOptionalString(Optional<String> optional) {
        return optional.map(String::toString).orElse("");
    }

    @TypeConverter
    public static Title toTitle(String id) {
        return TitleRepository.INSTANCE.loadTitle(id);
    }

    @TypeConverter
    public static String fromTitle(Title title) {
        return title.getId();
    }

    @TypeConverter
    public static Optional<Title> toOptionalTitle(String id) {
        return (id == null || id.isEmpty()) ? Optional.empty() : Optional.of(TitleRepository.INSTANCE.loadTitle(id));
    }

    @TypeConverter
    public static String fromOptionalTitle(Optional<Title> title) {
        return title.isPresent() ? title.get().getId() : "";
    }

    @TypeConverter
    public static IntentionStatus toIntentionStatus(String name) {
        return IntentionStatus.valueOf(name);
    }

    @TypeConverter
    public static String fromIntentionStatus(IntentionStatus intentionStatus) {
        return intentionStatus.name();
    }

    @TypeConverter
    public static IntentionType toIntentionType(String name) {
        return IntentionType.valueOf(name);
    }

    @TypeConverter
    public static String fromIntentionType(IntentionType intentionType) {
        return intentionType.name();
    }
}

File: app/src/main/java/com/ceaver/assin/database/Converters.kt
package com.ceaver.assin.database

import androidx.room.TypeConverter
import com.ceaver.assin.action.ActionType
import com.ceaver.assin.alerts.AlertType
import com.ceaver.assin.assets.AssetCategory
import com.ceaver.assin.intentions.IntentionStatus
import com.ceaver.assin.intentions.IntentionType
import java.math.BigDecimal
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import java.util.*

class Converters {

    @TypeConverter
    fun fromLocalDate(localDate: LocalDate?): Long? = localDate?.let { ChronoUnit.DAYS.between(LocalDate.MIN, localDate) }

    @TypeConverter
    fun toLocalDate(long: Long?): LocalDate? = long?.let { LocalDate.MIN.plusDays(it) }

    @TypeConverter
    fun fromBigDecimal(bigDecimal: BigDecimal?): String? = bigDecimal?.toPlainString()

    @TypeConverter
    fun toBigDecimal(string: String?): BigDecimal? = string?.toBigDecimalOrNull()

    @TypeConverter
    fun fromLocalDateTime(localDateTime: LocalDateTime?): Long? = localDateTime?.let { ChronoUnit.SECONDS.between(LocalDateTime.MIN, localDateTime) }

    @TypeConverter
    fun toLocalDateTime(long: Long?): LocalDateTime? = long?.let { LocalDateTime.MIN.plusSeconds(it) }

    @TypeConverter
    fun fromUuid(uuid: UUID?): String? = uuid?.toString()

    @TypeConverter
    fun toUuid(string: String?): UUID? = string?.let { UUID.fromString(string) }

    @TypeConverter
    fun fromAlertType(alertType: AlertType?): String? = alertType?.name

    @TypeConverter
    fun toAlertType(string: String?): AlertType? = string?.let { AlertType.valueOf(it) }

    @TypeConverter
    fun fromIntentionStatus(intentionStatus: IntentionStatus?): String? = intentionStatus?.name

    @TypeConverter
    fun toIntentionStatus(string: String?): IntentionStatus? = string?.let { IntentionStatus.valueOf(it) }

    @TypeConverter
    fun fromIntentionType(intentionType: IntentionType?): String? = intentionType?.name

    @TypeConverter
    fun toIntentionType(string: String?): IntentionType? = string?.let { IntentionType.valueOf(it) }

    @TypeConverter
    fun fromAssetCategory(assetCategory: AssetCategory?): String? = assetCategory?.name

    @TypeConverter
    fun toAssetCategory(string: String?): AssetCategory? = string?.let { AssetCategory.valueOf(it) }

    @TypeConverter
    fun fromActionType(assetType: ActionType?): String? = assetType?.name

    @TypeConverter
    fun toActionType(string: String?): ActionType? = string?.let { ActionType.valueOf(it) }
}